xarray/core/weighted.py

```diff
@@ -0,0 +1,307 @@
+
+    Parameters
+    ----------
+    dim : str or sequence of str, optional
+        Dimension(s) over which to apply the weighted `{fcn}`.
+    axis : int or sequence of int, optional
+        Axis(es) over which to apply the weighted `{fcn}`. Only one of the
+        'dim' and 'axis' arguments can be supplied. If neither are supplied,
+        then the weighted `{fcn}` is calculated over all axes.
+    skipna : bool, optional
+        If True, skip missing values (as marked by NaN). By default, only
+        skips missing values for float dtypes; other dtypes either do not
+        have a sentinel missing value (int) or skipna=True has not been
+        implemented (object, datetime64 or timedelta64).
+        Note: Missing values in the weights are replaced with 0 (i.e. no
+        weight).
+    keep_attrs : bool, optional
+        If True, the attributes (`attrs`) will be copied from the original
+        object to the new one.  If False (default), the new object will be
+        returned without attributes.
+    **kwargs : dict
+        Additional keyword arguments passed on to the appropriate array
+        function for calculating `{fcn}` on this object's data.
+
+    Returns
+    -------
+    reduced : {cls}
+        New {cls} object with weighted `{fcn}` applied to its data and
+        the indicated dimension(s) removed.
+    """
+
+_SUM_OF_WEIGHTS_DOCSTRING = """
+    Calculate the sum of weights, accounting for missing values
+
+    Parameters
+    ----------
+    dim : str or sequence of str, optional
+        Dimension(s) over which to sum the weights.
+    axis : int or sequence of int, optional
+        Axis(es) over which to sum the weights. Only one of the 'dim' and
+        'axis' arguments can be supplied. If neither are supplied, then
+        the weights are summed over all axes.
+
+    Returns
+    -------
+    reduced : {cls}
+        New {cls} object with the sum of the weights over the given dimension.
+
+
+    """
+
+
+# functions for weighted operations for one DataArray
+# NOTE: weights must not contain missing values (this is taken care of in the
+# DataArrayWeighted and DatasetWeighted cls)
+
+
+def _maybe_get_all_dims(
+    dims: Optional[Union[Hashable, Iterable[Hashable]]],
+    dims1: Tuple[Hashable, ...],
+    dims2: Tuple[Hashable, ...],
+):
+    """ the union of dims1 and dims2 if dims is None
+
+    `dims=None` behaves differently in `dot` and `sum`, so we have to apply
+    `dot` over the union of the dimensions
+
+    """
+
+    if dims is None:
+        dims = tuple(sorted(set(dims1) | set(dims2)))
+
+    return dims
+
+
+def _sum_of_weights(
+    da: "DataArray",
+    weights: "DataArray",
+    dim: Optional[Union[Hashable, Iterable[Hashable]]] = None,
+    axis=None,
+) -> "DataArray":
+    """ Calcualte the sum of weights, accounting for missing values """
+
+    # we need to mask DATA values that are nan; else the weights are wrong
+    mask = where(da.notnull(), 1, 0)  # binary mask
```

Comment:
```suggestion
    mask = da.isnull()
```
`bool` works with `dot` from my limited checks